[APM] Navigate to default transaction sample

dgieselaar opened the pull request:
Closes #33695, #33545.

If the distribution data changes, and the currently selected transaction sample is no longer available, we navigate to the default sample of the latest distribution data.

Previously, a displayed sample would "stick" if the new distribution data did not have a default sample.

## Summary

- Moves TransactionDetails/Distribution from a class component w/ lifecycle methods to a functional component w/ hooks. The assumption is that hooks are better at describing when things should change than lifecycle methods (my 2$: code is a little more complicated, but more robust as well).
- Basic idea here: on any distribution data change, check if the selected transaction sample is available in the data, if not, either: A) clear the sample when the new data has no default sample, B) select the default sample from the new data.
- In TransactionDetails, we're now stuck with two empty messages if for the given filters, there is no data. This PR removes the empty message from the Transaction sample component if there is no transaction, so only one message is displayed.

I'm looking for feedback on:

- Is this usage of hooks correct and in line with how we like to move forward? Especially the stuff around memoizing callbacks to be able to pass them as dependencies, and using ref for location (without it, we run into an infinite loop), seems complicated. Can we make it more straightforward without losing the robustness?
- Perhaps it's better to lift the redirect logic into TransactionDetails? I'm not sure why it's in Distribution, feels too deep, and perhaps things (like the stacked empty messages) are handled in a better way if that logic (and deciding what "empty" means) is in TransactionDetails.
- Any suggestions on what tests would need to be added?

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

dgieselaar committed:
[APM] Navigate to default transaction sample

If the distribution data changes, and the currently selected transaction sample is no longer available, we navigate to the default sample of the latest distribution data.

Previously, a displayed sample would "stick" if the new distribution data did not have a default sample.

Readability improvements

Use location primitives

Simplifies hook logic in TransactionDetails/Distribution by only passing the necessary primitives (pathname, hash, search) to the history.replace() call rather than the full location object. Removes the need for memoization w/ useRef too.

Fix lint error

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12977/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13009/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13090/)

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Distribution/index.tsx:
```diff
@@ -47,191 +53,226 @@ export function getFormattedBuckets(buckets: IBucket[], bucketSize: number) {
+  props: Props
+) => {
+  const {
+    location: sourceLocation,
```
Why rename to `sourceLocation`? Is that to avoid overwriting `window.location`? 

As an aside, I've noticed an issue where we (by mistake) access `window.location` when writing `location` - I think we can solve this with the eslitn rule `no-restricted-globals`

sqren replied:
Oh, I see. You declare `location` below from the ref.

dgieselaar replied:
@sqren I think that rule is already in use:

![image](https://user-images.githubusercontent.com/352732/57403378-de202000-71d9-11e9-85e3-9bc129b8c636.png)

sqren replied:
Interesting, I only get those warnings in js-files but not in ts.

dgieselaar commented:
@sqren I've addressed your feedback about location. I now pass primitives rather than the complete object, so I no longer need `useRef()`, or the `isEqual()` check in `redirectToDefaultSample`. Looks much better, thanks 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13315/)

sqren commented:
@dgieselaar Awesome. Will take a look.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13522/)

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Distribution/index.tsx:
```diff
@@ -47,191 +48,212 @@ export function getFormattedBuckets(buckets: IBucket[], bucketSize: number) {
+            traceId: defaultSample.traceId,
+            transactionId: defaultSample.transactionId
+          }
+        : {};
```
Would this work?

```js
const defaultSample =
  distribution && distribution.defaultSample ? distribution.defaultSample : {};
// ...

history.replace({
  pathname,
  hash,
  search: fromQuery({
    ...omit(parsedQueryParams, 'transactionId', 'traceId'),
    ...defaultSample
  })
});
```

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Distribution/index.tsx:
```diff
@@ -47,191 +48,212 @@ export function getFormattedBuckets(buckets: IBucket[], bucketSize: number) {
+        })
+      });
+    },
+    [distribution, pathname, hash, search]
```
This might be correct, but I'm wondering:
if we care about `pathname`, `hash` and `search`, why not pass in `[distribution, location]`?

On the other hand: do we even care about location changing? Isn't it only `[distribution, transactionId, traceId]` we care about?

dgieselaar replied:
I'll check again, but I think this is because if we pass location as a dependency to `redirectToDefaultSample`, it's identity changes every time location changes (even though its _value_ has not changed. Because we pass the callback to the `useEffect` hook as a dependency, that hook is being executed too often as well. I might be able to clean it up though.

dgieselaar replied:
Yeah, the renderer gets stuck in a loop. The initial render triggers a redirect to the default sample of the selected transaction distribution. Because of that redirect, the (identity of the) `location` object changes. That will trigger the redirect hook to be called on successive renders too, ad infinitum. However, because the values of the `location` object do not change, we can prevent this by using the primitives.

sqren replied:
I see. What do you think about getting location from the `history.location` singleton? `history.location` is a mutable object, so cannot be used to detect changes but that is also not needed here (on the contrary actually). So something like:

```js
  const redirectToDefaultSample = useCallback(
    () => {    
      const parsedQueryParams = toQuery(history.location.search);
      const defaultSample =  idx(distribution, _ => _.defaultSample) || {};

      history.replace({
        ...history.location,
        search: fromQuery({
          ...omit(parsedQueryParams, 'transactionId', 'traceId'),
          ...defaultSample
        })
      });
    },
    [distribution]
  );
```

Of course this only works if all we care about are changes to `distribution`.

sorenlouv reviewed (approved):
LGTM

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15472/)

dgieselaar committed:
Remove unused translations

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15475/)

dgieselaar committed:
Simplify redirectToDefaultSample callback

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15487/)

dgieselaar committed:
Use history.location instead of location prop

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15491/)
